Support for HAFS UFSATM-HYCOM and DATM-HYCOM coupling.

danrosen25 opened the pull request:
### Description of changes
Added mapping type for UFSATM-HYCOM coupling. New mapping type fills in destination points with 9.99e20, a physically unrealistic value, before executing bilinear regridding. Then bilinear regridding does not modify unmapped points. This allows destination component to identify unmapped points.

Small fixes were made to check FieldBundleIsCreated before accessing data pointer from FB. I believe this is due to a domain with a sparse decomposition.

HAFS field coupling configuration was modified/added to esmFldsExchange_hafs_mod.F90. There are now two configurations.  system_type=CDEPS configures DATM-Active_Ocean coupling and includes flux calculation in the mediator.  system_type=UFS configures Active_Atmosphere-Active_Ocean coupling and does not compute fluxes.  I do have an Issue opened for this topic but the issue is only temporarily addressed here by adding the system_type to esmFldsExchange_hafs_mod.F90.
https://github.com/ESCOMP/CMEPS/issues/92

This PR also aims to remove internal PIO from CMEPS completely and move the shared code to another folder (we removed the nems from the directory structure). The PIO initialization is in the NEMS side. @DeniseWorthen you need to use following branch in the PR to work under S2S (https://github.com/hafs-community/NEMS/pull/2) because NEMS has additional mods that is required for PIO initialization.

### Specific notes

Contributors other than yourself, if any:
@uturuncoglu - Worked on the DATM-Active_Ocean coupling for HAFS including internal PIO removal
@danrosen25 - Worked on the Active_Atmosphere-Active_Ocean coupling for HAFS including new mapping type

CMEPS Issues Fixed (include github issue #):
Addressed this issue by creating 'mapfillv_bilnr'
https://github.com/ESCOMP/CMEPS/issues/91

Are changes expected to change answers?
 - [x] bit for bit
 - [ ] different at roundoff level
 - [ ] more substantial 

Any User Interface Changes (namelist or namelist defaults changes)?
 - [ ] Yes
 - [x] No

Testing performed if application target is CESM:(either UFS-S2S or CESM testing is required):
- [x] (recommended) CIME_DRIVER=nuopc scripts_regression_tests.py
   - machines: Cheyenne, Intel 19.0.5 and MPT 2.15
   - details (e.g. failed tests): Some tests are failed but after following CDEPS PR (https://github.com/ESCOMP/CDEPS/pull/47) all of them running now.
- [x] (recommended) CESM testlist_drv.xml
   - machines and compilers: Cheyenne, Intel 19.0.5 and MPT 2.15
   - details (e.g. failed tests): There are no failed tests expect following computation and memory increases
   ```
    FAIL ERS_Vnuopc_Ln9_N3.f19_g17_rx1.A.cheyenne_intel TPUTCOMP Error: TPUTCOMP: Computation time increase > 25% from baseline
    FAIL SMS_Vnuopc_Ld1_N3.f19_g17_rx1.A.cheyenne_intel TPUTCOMP Error: TPUTCOMP: Computation time increase > 25% from baseline
    FAIL SMS_Vnuopc_Ld5.T62_t061.GMOM.cheyenne_intel MEMCOMP Error: Memory usage increase > 10% from baseline
   ```
- [ ] (optional) CESM prealpha test
   - machines and compilers
   - details (e.g. failed tests):
- [ ] (other) please described in detail
   - machines and compilers
   - details (e.g. failed tests): 

Testing performed if application target is UFS-coupled:
- [ x] (recommended) UFS-coupled testing
   - description: all coupled and datm tests pass
   - details (e.g. failed tests):

Testing performed if application target is UFS-HAFS:
- [x] (recommended) UFS-HAFS testing
   - description: Testing conducted with UFSATM-HYCOM coupling and DATM-HYCOM coupling on Hera and Orion
   - details (e.g. failed tests): Testing will not produce bit-for-bit results when compared to UFSATM-HYCOM coupled through NUOPC Connector. This is expected due to term ordering differences in the SMM factor generation for a grid converted to a mesh. ESMF has a branch of work that can be used to synchronize the term ordering https://github.com/esmf-org/esmf/tree/gtom_bfb_debug

Hashes used for testing:
- [x] CESM:
  - repository to check out: https://github.com/ESCOMP/CESM.git
  - branch: nuopc_dev
  - hash: c019fc2
- [x] UFS-coupled, then umbrella repository to check out and associated hash:
  - repository to check out:https://github.com/ufs-community/ufs-weather-model
  - branch:develop, then point CMEPS to support/HAFS branch and modify CMEPS/CMakelist.txt
  - hash:62991a1
- [x] UFS-HAFS, then umbrella repository to check out and associated hash:
  - repository to check out: https://github.com/hafs-community/HAFS
  - branch: feature/hafs_couplehycom_cmeps
  - hash: 57aa33baa97670bdbf6ca100cabe65f0ee71ba50

danrosen25 commented:
@uturuncoglu  
Build failed due to the internal PIO removal. Does the build-cmeps test script need to be updated to account for the PIO change?

uturuncoglu commented:
@danrosen25 if you are talking about the GitHub action failure this is expected. We need to fix it. @jedwards4b do you want me to look at?

uturuncoglu commented:
@mvertens @DeniseWorthen I'll test this branch with CESM and add the details to PR.

mvertens commented:
@uturuncoglu - thank you for offering to test with CESM. It was not clear to me from the changes in med_map_mod that this would not effect the cesm results - or the UFS results.

uturuncoglu commented:
@mvertens the changes in med_map_mod probably does not affect other applications because @danrosen25 introduced a new interpolation and I put some fix for the masking in the data component and they are isolated for HAFS app. Eventually, the definition of source and destination mask values need to be revisited to make this part of code more clean in CMEPS side. Currently, every application (or coupling mode) has its own way. Anyway, let me run the full CESM test suite and see what happens. If we see any answer change or etc. we could log at more detailed. Which baseline do I need to use for it?

mvertens commented:
That sounds like a great idea. Thanks.

On Thu, Jan 21, 2021 at 2:53 PM Ufuk Turunçoğlu <notifications@github.com>
wrote:

> @mvertens <https://github.com/mvertens> the changes in med_map_mod
> probably does not affect other applications because @danrosen25
> <https://github.com/danrosen25> introduced a new interpolation and I put
> some fix for the masking in the data component and they are isolated for
> HAFS app. Eventually, the definition of source and destination mask values
> need to be revisited to make this part of code more clean in CMEPS side.
> Currently, every application (or coupling mode) has its own way. Anyway,
> let me run the full CESM test suite and see what happens. If we see any
> answer change or etc. we could log at more detailed. Which baseline do I
> need to use for it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CMEPS/pull/154#issuecomment-764965181>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB4XCEZIWT7OD5S3FWUIFSDS3CO3ZANCNFSM4WNO6ORA>
> .
>


-- 
Mariana Vertenstein
CESM Software Engineering Group Head
National Center for Atmospheric Research
Boulder, Colorado
Office 303-497-1349
Email: mvertens@ucar.edu

DeniseWorthen commented:
I'm not able to build this in ufs-weather;  I'm getting a error : 

```
-- Configuring done
CMake Error at CMEPS-interface/CMakeLists.txt:73 (add_library):
  Cannot find source file:

    CMEPS/nems/util/shr_abort_mod.F90

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
  .hpp .hxx .in .txx


CMake Error at CMEPS-interface/CMakeLists.txt:73 (add_library):
  No SOURCES given to target: cmeps


CMake Generate step failed.  Build files cannot be regenerated correctly.

```

I clone my current ufs-weather fork and pointed to the support/HAFS branch ( /glade/work/worthen/ufs_testhafs).

uturuncoglu commented:
@DeniseWorthen I think you also need to update UFS model itself, at least the CMEPS-interface/CMakeLists.txt file to test with S2S. Could you replace your CMEPS-interface/CMakeLists.txt file with the following and try again,

https://github.com/hafs-community/ufs-weather-model/blob/feature/hafs_couplehycom/CMEPS-interface/CMakeLists.txt

Eventually, all those model level changes need to be synced. Let me know if you have any other problem.

danrosen25 commented:
Hi @DeniseWorthen 

The files in CMEPS/nems/util/ were moved https://github.com/ESCOMP/CMEPS/pull/154/commits/a823ee5423841236233e7279e4b90a7e43553f93

You'll have to update the _nems_util_files list in CMEPS-interface/CMakeLists.txt 

list(APPEND _nems_util_files
  CMEPS/util/shr_abort_mod.F90
  CMEPS/util/shr_log_mod.F90
  CMEPS/util/shr_pio_mod.F90
  CMEPS/util/shr_sys_mod.F90
  CMEPS/util/shr_flux_mod.F90
  CMEPS/util/shr_mpi_mod.F90
  CMEPS/util/glc_elevclass_mod.F90
  CMEPS/util/shr_mem_mod.F90
  CMEPS/util/shr_kind_mod.F90
  CMEPS/util/perf_mod.F90
  CMEPS/util/shr_const_mod.F90)

jedwards4b commented:
@danrosen25 If you merge the latest master to this PR the extbuild should now pass.

danrosen25 commented:
@jedwards4b Is it okay to remove the empty .gitmodules file?

jedwards4b commented:
yes, thanks

danrosen25 commented:
@mvertens I submitted an issue for the hard coded masking by coupling_mode back in August https://github.com/ESCOMP/CMEPS/issues/93

DeniseWorthen commented:
@uturuncoglu @danrosen25 I've gotten it to build now in ufs-weather. I'll run the ufs-coupled and datm tests.

jedwards4b commented on mediator/med.F90:
```diff
@@ -835,14 +851,14 @@ subroutine InitializeIPDv03p1(gcomp, importState, exportState, clock, rc)
-       write(logunit,*) "wav_present="//trim(wav_present)
-       write(logunit,*) "glc_present="//trim(glc_present)
-       write(logunit,*) "med_present="//trim(med_present)
+       write(logunit,*) "atm_present = "//trim(atm_present)//" - atm_name = "//trim(atm_name)
```
What is atm_name if atm_present is false?   maybe instead of this you could have
if(atm_present) write(logunit,*) "atm_name = "//trim(atm_name)

danrosen25 replied:
Strings are initialized to blank values. Is this okay? See line [81](https://github.com/ESCOMP/CMEPS/pull/154/files/45eb748ee19cfed10575a2ce2c7ac72d618db43c#diff-fd9e77eeb56b02d6d6a3b93a1e5d47bb1d6f83b3557a22ba1c17dca9a0d35ee4R81) in mediator/esmFlds.F90.

danrosen25 replied:
Option 2: We can also set the atm_name to something like "NA" if you don't like " " here:
[https://github.com/hafs-community/CMEPS/blob/support/HAFS/mediator/med.F90#L778](https://github.com/hafs-community/CMEPS/blob/support/HAFS/mediator/med.F90#L778)

If we do add if (atm_present) the code won't be as concise, the code for each component will look like...
```
if (atm_present) then
   write(logunit,*) "atm_present = "//trim(atm_present)//" - atm_name = "//trim(atm_name)
else
   write(logunit,*) "atm_present = "//trim(atm_present)
endif
```

jedwards4b replied:
why write anything at all if it's false? 
if (atm_present) write(logunit,*) "atm_name = "//trim(atm_name)

is all you need.

danrosen25 replied:
med.F90 currently writes whether a component is present or not.
https://github.com/ESCOMP/CMEPS/blob/master/mediator/med.F90#L836

jedwards4b replied:
it's not worth the fight

jedwards4b commented on mediator/med_map_mod.F90:
```diff
@@ -272,6 +274,17 @@ subroutine med_map_routehandles_initfrom_field(n1, n2, fldsrc, flddst, mapindex,
        if (n2 == compocn .or. n2 == compice) dstMaskValue = 0
+       if (n1 == compatm .and. n2 == compocn) then
+          if (trim(atm_name).eq.'datm') then
+             srcMaskValue = ispval_mask
```
This was already set in line 274, no need to do it again.

danrosen25 replied:
It was only in there for readability, we can remove it. We can get really technical and remove the trim because the fortran standard will pad the end of a string with " " until the two strings are of equal length. Are you okay with this change...
```
if (n1 == compatm .and. n2 == compocn) then
   if (atm_name.ne.'datm') then
      srcMaskValue = 1
   endif 
```

jedwards4b replied:
I'm okay but for consistency I think the trim should be included.

jedwards4b commented on mediator/med_methods_mod.F90:
```diff
@@ -737,11 +738,16 @@ subroutine med_methods_FB_reset(FB, value, rc)
+       if (chkerr(rc,__LINE__,u_FILE_u)) return
+    else
+       fieldCount=0
+       allocate(lfieldnamelist(fieldCount))
```
I think that some compilers will throw an error on allocate with count=0

danrosen25 replied:
Section 2.4.5 of the F90 standard says that arrays can have zero size so I hope that all compilers can handle allocate with count=0.
[F90 Standard](https://wg5-fortran.org/N001-N1100/N692.pdf)
Do you still want me to change this? It just means that I have to check before the deallocation and make sure that the array isn't used at all.

jedwards4b replied:
I remember having problems with this but don't remember which compiler or how far back it was, you can leave this one.

danrosen25 commented:
@uturuncoglu The CMEPS/CMakeLists.txt and CMEPS/mediator/CMakeLists.txt files haven't been updated to account for the nems/util move. Can you fix this?

uturuncoglu commented:
@danrosen25 i am workin on fixing Cmake files. It will be ready soon but I same that our branch has nems/ folder now. Is it because of merging the branch with master to fix the extbuild issue?

uturuncoglu commented:
@danrosen25 i fixed the cmake issue but I could not push those changes to support/HAFS directly because it is protected I think or I have no rights to do it. So, I am getting `error: unable to push to unqualified destination: fork/support/HAFS`. I might create a branch in the fork and make it there and you could merge that branch with support/HAFS. What do you think?

uturuncoglu commented:
@mvertens @jedwards4b @danrosen25 I updated PR description and add results of full CESM test suite. I am also running scripts_regression_tests.py and once it is done I'll add it too.

DeniseWorthen commented on mediator/med_fraction_mod.F90:
```diff
@@ -299,7 +300,23 @@ subroutine med_fraction_init(gcomp, rc)
        call fldbun_getdata1d(is_local%wrap%FBFrac(compocn), 'ofrac', ofrac, rc)
        if (ChkErr(rc,__LINE__,u_FILE_u)) return
-       ofrac(:) = So_omask(:)
+       if (associated(So_omask)) then
```
I don't quite understand why this is required. Won't the call to fldbun_getdata1d return an error if So_omask doesn't exist in FBImp(compocn,compocn)? 

Why isn't a similar change required at LN 249 (setting Si_imask and ifrac)?

danrosen25 replied:
Hi @DeniseWorthen . It doesn't return an error and you don't want it to return an error. I found this because the ocean decomposition is sparse, meaning there are PETs that don't allocate any ocean cells. In this case FBFrac(compocn) returns a null pointer. It can be fixed another way through allocating zero sized objects but that fix has to go deeper into the code.

A similar change should be made for ifrac but HAFS doesn't have an ice component and therefore it didn't crash. I can replicate this code for the other fractions but I won't be able to test the change.

danrosen25 replied:
This work is proving me a little more trouble. In some cases the logic is complex. Would you approve a change that looked more like this? In this case it simplifies the check but takes care of my unique case.
```
       if (associated(ofrac)) then
          ofrac(:) = So_omask(:)
       endi
```

danrosen25 commented:
@uturuncoglu 
@jedwards4b moved nems/util and fixed the CMakeLists.txt files and that's now on the master. I didn't notice the conflicts related to PIO. Can you decide on how to handle the PIO related conflicts?

### Makefile ###
```
$(LIBRARY_UTIL): .FORCE
        cd util ;\
<<<<<<< support/HAFS
        exec $(MAKE) PIO_INC=$(PIO_INC)
=======
        exec $(MAKE)
>>>>>>> master
```

### mediator/Makefile ###
```
%.o: %.F90
<<<<<<< support/HAFS
        $(ESMF_F90COMPILER) -c $(ESMF_F90COMPILEOPTS) $(ESMF_F90COMPILEPATHS) $(CPPDEFS) -I${PIO_INC} -I../util $*.F90
=======
        $(ESMF_F90COMPILER) -c $(ESMF_F90COMPILEOPTS) $(ESMF_F90COMPILEPATHS) $(CPPDEFS) -I$(PIO_ROOT)/include -I../util $*.F90
>>>>>>> master
```

uturuncoglu commented:
@danrosen25 okay, I'll look at the conflict also.

uturuncoglu commented:
@danrosen25 it seems that @jedwards4b also fixe the issue related with the cmake build (removing nems directories). I just only push a minor fix in attribute get call. The variable was not defined, now it should be fine.

uturuncoglu commented:
@jedwards4b I run the regression test and I have couple of assertion error which I don't know why. The log is in `/glade/u/home/turuncu/NCAR/CESM_test/cime/scripts/tests/log`. If you don't mind could you check it. Thanks.

jedwards4b commented:
You need to update (or remove) the mizuroute component, it has a bad
config_archive.xml file

On Fri, Jan 22, 2021 at 12:58 PM Ufuk Turunçoğlu <notifications@github.com>
wrote:

> @jedwards4b <https://github.com/jedwards4b> I run the regression test and
> I have couple of assertion error which I don't know why. The log is in
> /glade/u/home/turuncu/NCAR/CESM_test/cime/scripts/tests/log. If you don't
> mind could you check it. Thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CMEPS/pull/154#issuecomment-765650565>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABOXUGEM55AZ7TLZBTUIRDDS3HKFPANCNFSM4WNO6ORA>
> .
>


-- 
Jim Edwards

CESM Software Engineer
National Center for Atmospheric Research
Boulder, CO

uturuncoglu commented:
@danrosen25 there is still problem in the `esmFldsExchange_hafs_attr` call. It was using value variable, which were not defined and I changed it to cvalue but still failing like following. 

```
/home/runner/work/CMEPS/CMEPS/mediator/esmFldsExchange_hafs_mod.F90:944:38:
       specialStringList=(/"off","low","high","max"/), &
Error: Different CHARACTER lengths (3/4) in array constructor at (1)
make[2]: *** [mediator/CMakeFiles/cmeps.dir/esmFldsExchange_hafs_mod.F90.o] Error 1
```

Since you add this part, could you fix it?

uturuncoglu commented:
@jedwards4b it seems it is using `mizuRoute` branch. Do I ned to use the head of this branch?

mvertens commented:
Please talk to Erik about this regarding mizuroute. They are doing active
development with this.

On Fri, Jan 22, 2021 at 1:08 PM Ufuk Turunçoğlu <notifications@github.com>
wrote:

> @jedwards4b <https://github.com/jedwards4b> it seems it is using mizuRoute
> branch. Do I ned to use the head of this branch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CMEPS/pull/154#issuecomment-765655570>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB4XCE3QYROHRHSTWZS6TK3S3HLLHANCNFSM4WNO6ORA>
> .
>


-- 
Mariana Vertenstein
CESM Software Engineering Group Head
National Center for Atmospheric Research
Boulder, Colorado
Office 303-497-1349
Email: mvertens@ucar.edu

jedwards4b commented:
They've already fixed the issue in a mizoroute update - but you can just
delete that component for your purposes.

On Fri, Jan 22, 2021 at 1:18 PM mvertens <notifications@github.com> wrote:

> Please talk to Erik about this regarding mizuroute. They are doing active
> development with this.
>
> On Fri, Jan 22, 2021 at 1:08 PM Ufuk Turunçoğlu <notifications@github.com>
> wrote:
>
> > @jedwards4b <https://github.com/jedwards4b> it seems it is using
> mizuRoute
> > branch. Do I ned to use the head of this branch?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/ESCOMP/CMEPS/pull/154#issuecomment-765655570>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AB4XCE3QYROHRHSTWZS6TK3S3HLLHANCNFSM4WNO6ORA
> >
> > .
> >
>
>
> --
> Mariana Vertenstein
> CESM Software Engineering Group Head
> National Center for Atmospheric Research
> Boulder, Colorado
> Office 303-497-1349
> Email: mvertens@ucar.edu
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CMEPS/pull/154#issuecomment-765660499>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABOXUGEIU2WAN2OIBUENHC3S3HMTDANCNFSM4WNO6ORA>
> .
>


-- 
Jim Edwards

CESM Software Engineer
National Center for Atmospheric Research
Boulder, CO

uturuncoglu commented:
@jedwards4b you mean deleting misroute component and run regression test again? Sorry for asking too much question but it was not clear for me. I just want to be sure.

jedwards4b commented:
yes either delete or update the mizoroute component

On Fri, Jan 22, 2021 at 1:25 PM Ufuk Turunçoğlu <notifications@github.com>
wrote:

> @jedwards4b <https://github.com/jedwards4b> you mean deleting misroute
> component and run regression test again? Sorry for asking too much question
> but it was not clear for me. I just want to be sure.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CMEPS/pull/154#issuecomment-765663623>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABOXUGHM32V3D76HKILNMBDS3HNMZANCNFSM4WNO6ORA>
> .
>


-- 
Jim Edwards

CESM Software Engineer
National Center for Atmospheric Research
Boulder, CO

uturuncoglu commented:
@jedwards4b i deleted  mizoroute component and run the regression tests again, I have still failed tests that I have no idea why. The log file is in `/glade/u/home/turuncu/NCAR/CESM_test/cime/scripts/tests/log2`

jedwards4b commented:
You're failing A cases.   Here is a PET file with the error:
 /glade/scratch/turuncu/scripts_regression_test.20210122_150316/PET_Ln9_P4.f19_f19.A.cheyenne_intel.fake_testing_only_20210122_162825/run/PET0.ESMF_LogFile

On Fri, Jan 22, 2021 at 11:34 PM Ufuk Turunçoğlu <notifications@github.com>
wrote:

> @jedwards4b <https://github.com/jedwards4b> i deleted mizoroute component
> and run the regression tests again, I have still failed tests that I have
> no idea why. The log file is in
> /glade/u/home/turuncu/NCAR/CESM_test/cime/scripts/tests/log2
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CMEPS/pull/154#issuecomment-765877647>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABOXUGDURDRNS3FLPGGQSEDS3JUWJANCNFSM4WNO6ORA>
> .
>


-- 
Jim Edwards

CESM Software Engineer
National Center for Atmospheric Research
Boulder, CO

uturuncoglu commented:
@jedwards4b i am not sure about what do I need to do. It seems that ESMF throws error about its dimensions when reading `/glade/p/cesmdata/cseg/inputdata/share/domains/domain.ocn.1.9x2.5_gx1v6_090403.nc`. It seems that this version of app uses ESMF 8.1.0b41. Did we test the CESM app with this version of ESMF?

jedwards4b commented:
We've been using b41 for a few weeks without issue.

uturuncoglu commented:
@jedwards4b okay. it is good to know. @danrosen25 do you have any idea about the issue?

jedwards4b commented:
Having said that it appears to be failing on master as well, i suspect the problem is in cdeps.

jedwards4b commented:
@mvertens I think you could probably fix this quickly,  this file is being used as meshfile_mask, but I think it's the wrong file isn't it?

uturuncoglu commented:
@mvertens i am not sure. maybe the updated CMEPS tires to use some feature which is not supported with the current CESM. Let me try to downgrade the CDEPS to be sure that it is not the issue and if pass, we could look at CMEPS and CIME side.

mvertens commented:
Please do not downgrade cdeps until I have a chance to test this. These
tests passed fine in my previous commit.

On Mon, Jan 25, 2021 at 10:47 AM Ufuk Turunçoğlu <notifications@github.com>
wrote:

> @mvertens <https://github.com/mvertens> i am not sure. maybe the updated
> CMEPS tires to use some feature which is not supported with the current
> CESM. Let me try to downgrade the CDEPS to be sure that it is not the issue
> and if pass, we could look at CMEPS and CIME side.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CMEPS/pull/154#issuecomment-766995207>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB4XCE6COCLZ7OHJJGQVFZDS3WVE3ANCNFSM4WNO6ORA>
> .
>


-- 
Mariana Vertenstein
CESM Software Engineering Group Head
National Center for Atmospheric Research
Boulder, Colorado
Office 303-497-1349
Email: mvertens@ucar.edu

mvertens commented:
Are you using the latest nuopc_dev hash to do this test?

On Mon, Jan 25, 2021 at 11:13 AM Mariana Vertenstein <mvertens@ucar.edu>
wrote:

> Please do not downgrade cdeps until I have a chance to test this. These
> tests passed fine in my previous commit.
>
> On Mon, Jan 25, 2021 at 10:47 AM Ufuk Turunçoğlu <notifications@github.com>
> wrote:
>
>> @mvertens <https://github.com/mvertens> i am not sure. maybe the updated
>> CMEPS tires to use some feature which is not supported with the current
>> CESM. Let me try to downgrade the CDEPS to be sure that it is not the issue
>> and if pass, we could look at CMEPS and CIME side.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/ESCOMP/CMEPS/pull/154#issuecomment-766995207>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AB4XCE6COCLZ7OHJJGQVFZDS3WVE3ANCNFSM4WNO6ORA>
>> .
>>
>
>
> --
> Mariana Vertenstein
> CESM Software Engineering Group Head
> National Center for Atmospheric Research
> Boulder, Colorado
> Office 303-497-1349
> Email: mvertens@ucar.edu
>


-- 
Mariana Vertenstein
CESM Software Engineering Group Head
National Center for Atmospheric Research
Boulder, Colorado
Office 303-497-1349
Email: mvertens@ucar.edu

jedwards4b commented:
These tests use master

uturuncoglu commented:
@mvertens yes, I am using CDEPS master. BTW, is `PET_Ln9_P4.f19_f19.A.cheyenne_intel.fake_testing_only` test same with `PET_Ln9_P4.f19_f19.A.cheyenne_intel`. I am asking because I could not create a test using `./create_test PET_Ln9_P4.f19_f19.A.cheyenne_intel.fake_testing_only`but `./create_test PET_Ln9_P4.f19_f19.A.cheyenne_intel` is working. @jedwards4b I did not downgrade CDEPS. BTW, `PET_Ln9_P4.f19_f19.A.cheyenne_intel` is passing with my PR also. So, I am not sure that I am testing same test that was failing. @jedwards4b let me know what do you think?

jedwards4b commented:
you need to add Vnuopc to your test to test CDEPS

uturuncoglu commented:
@jedwards4b okay, let me try again. Thanks.

mvertens commented:
I have a fix for this in CDEPS and am happy to push back to CDEPS master if
that would help - or issue a PR. Let me know how you want to proceed.

On Mon, Jan 25, 2021 at 12:10 PM jedwards4b <notifications@github.com>
wrote:

> you need to add Vnuopc to your test to test CDEPS
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CMEPS/pull/154#issuecomment-767045704>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB4XCE25CZR7LTWNEP4H67LS3W6Y3ANCNFSM4WNO6ORA>
> .
>


-- 
Mariana Vertenstein
CESM Software Engineering Group Head
National Center for Atmospheric Research
Boulder, Colorado
Office 303-497-1349
Email: mvertens@ucar.edu

uturuncoglu commented:
@mvertens @jedwards4b thanks for quick fix. I am testing it again.

uturuncoglu commented:
@mvertens @jedwards4b all the tests are passed after recent fix that Mariana put in CDEPS side. You could see the full log in the following path `/glade/u/home/turuncu/NCAR/CESM_test/cime/scripts/tests/log3`

jedwards4b commented:
You still have a fail in the github build.  It's a real error that needs to be fixed:
/home/runner/work/CMEPS/CMEPS/mediator/esmFldsExchange_hafs_mod.F90:944:38:

       specialStringList=(/"off","low","high","max"/), &

uturuncoglu commented:
@jedwards4b yes, I know. I'll look at it. CESM tests and regression tests are compiled without any problem. I think it is related with GNU. I'll fix it soon.

uturuncoglu commented:
@jedwards4b let's see what happens. I hope it will pass this time.

danrosen25 commented:
@jedwards4b @DeniseWorthen 
Hi Jim and Denise,
I'm going to make all the changes as suggested. I apologize for the back and forth as I was just waiting for clarity. I will resubmit a re-review shortly.

uturuncoglu commented:
@danrosen25 I need to check this version of PR with data components especially modifications related with the masking. I'll let you know about it.

danrosen25 commented:
Hi @jedwards4b @DeniseWorthen 
I believe I addressed all the code review comments left in the latest commit.
https://github.com/ESCOMP/CMEPS/pull/154/commits/ab3fd304ec2720d6a7bfddc2ecc1d5adb37b4241

I apologize for the string array mismatching length issue. That is something that depends on the compiler and I didn't notice it.

DeniseWorthen commented:
I will run a sub-set of our tests just to confirm. I don't expect any negative impacts. Thanks.

DeniseWorthen reviewed (approved):
Looks good.

jedwards4b commented on mediator/med.F90:
```diff
@@ -835,14 +851,46 @@ subroutine InitializeIPDv03p1(gcomp, importState, exportState, clock, rc)
 
     if (mastertask) then
        write(logunit,*)
```
I really think that you missed my point here.  If the component is present print the name of the component, if no component is present don't print anything at all.  What you have is worse than what you are replacing.

danrosen25 replied:
No, I understand. Thanks.

uturuncoglu commented:
@danrosen25 i tested this version of brach and it works without any problem in the data configurations.

mvertens commented:
*@uturuncoglu -* thanks for your testing.

On Wed, Jan 27, 2021 at 10:13 AM Ufuk Turunçoğlu <notifications@github.com>
wrote:

> *@uturuncoglu* approved this pull request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CMEPS/pull/154#pullrequestreview-577570272>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB4XCE566AW6ZBGCUBNUHN3S4BCTHANCNFSM4WNO6ORA>
> .
>


-- 
Mariana Vertenstein
CESM Software Engineering Group Head
National Center for Atmospheric Research
Boulder, Colorado
Office 303-497-1349
Email: mvertens@ucar.edu

danrosen25 commented:
I apologize for another change. I reviewed the PR code changes one more time and found another allocate size zero. I also modified the previous allocate size zero fix to match the logic used for this one for consistency.  It won't have an effect on any of the previous runs.
https://github.com/ESCOMP/CMEPS/pull/154/commits/19ff711593282f1e20afbbb8b4bda944ed058344
